### ISSUE TITLE
Safari onAuthStateChanged stuck when multiple tabs open

### ISSUE THREAD

**Issue by ReinoutStevens** [firebase/firebase-js-sdk#8019]
### Operating System

iOS

### Browser Version

Safari

### Firebase SDK Version

10.7.2

### Firebase SDK Product:

Auth

### Describe your project's tooling

NextJS app with Firebase loading as a dynamic component / separate chunk

### Describe the problem

There is a problem with the `onAuthStateChanged` event callback in Safari on iOS ( or just with authentication in general ). The callback is not always invoked when the app has started, and as a result, the user is stuck while our app is waiting for firebase to finish. 
More specifically, the callback can also be invoked unreasonably late, as if firebase is stuck somewhere.

Initially, we could not reproduce this reliably, and have only seen it ourselves a couple of times.
As such, we have added an analytics event in the `onAuthStateChanged` callback to trigger when and how often firebase loads.
We just found out that the behavior can be triggered by having multiple tabs open to the same site, and after several hours, opening a new tab will get stuck.

Not loading:

| Browser | Sessions Loading | Sessions Not Loading | Ratio |
| ------- | ---------------- | -------------------- | ----- |
| Chrome  | 16311503 | 4512 | 0.027 % |
| Firefox | 734236 | 59 | 0.008% |
| Safari | 2053968 | 22709 | 1% |

Delayed loading:

For sessions in which firebase loads we looked at the different percentiles, median and average.
Timings are in ms

| Browser | avg time | 5pct | 25pct | median | 75pct | 95pct |
| ------- | -------- | ---- | ----- | ------ | ----- | ----- |
| Chrome  | 1763     | 2 | 10 |	133 | 497 | 2239 |
| Firefox | 3097     | 1 | 3 | 9 | 134 | 911 |
| Safari | 427408    | 1 | 7 | 77 | 371 | 16964 |

Yes, that average for Safari is correct. For a small subset of users the onAuthStateChanged callback takes an insanely long amount of time to get called.

We have reworked our application a bit to try to not give these users an awful experience.
For users where we suspect that firebase is not loading, but that still stick around long enough, we notice that firebase can eventually load after some time.

Timings are in seconds

| Browser | avg time | 5pct | 25pct | median | 75pct | 95pct |
| ------- | -------- | ---- | ----- | ------ | ----- | ----- |
| Chrome  | 197	     | 0    | 5     |18  | 41 | 163 |
| Firefox | 339      | 0 |9 | 29 | 76 | 259 |
| Safari | 7578 | 0 |31 | 235 |1700 | 38987 |

Chrome and Firefox timings can be explained due to slow devices or networks ( as we're quite a heavy site in terms of downloading assets ), but Safari has other problems.

### Steps and code to reproduce issue

We recently ( read, last night ) finally found out that you can get this behavior by having multiple tabs open of site for a couple of hours ( we managed to get it after waiting 2h, we don't know if you can get it sooner ), and then opening a new tab to the site. This tab will be stuck. Refreshing any of the other tabs will also result in them being stuck. Closing some of the tabs will result in a stuck tab to become unstuck.

We haven't looked deeper into how long tabs need to be open, or whether it's a specific tab you need to close.
In case it matters, we've had it happen on a device where we never logged in on firebase, and a device where we had previously logged in. In both cases after a couple of refreshes firebase loaded just fine.

The code boils down to:

```
const [userLoading, setUserLoading] = React.useState(false);

useEffect(() => {
    const cleanup = onAuthStateChanged(firebaseAuth, (newUser) => {
       // callback gets never invoked
       setUserLoading(false);
   };
  return cleanup;
}
```

We have tried other implementations using `authStateReady` ( which under the hood also use `onAuthStateChanged` ), where the same problem occurs.

Related issues:
https://github.com/firebase/firebase-js-sdk/issues/2167

**Comment by ReinoutStevens** [comment]
I dislike bumping issues by adding a comment, but this issue is still affecting our site. We're adding notices to our users when firebase doesnt load within a certain timeframe to kill other tabs, but this is a really abysmal user experience.

**Comment by jacobgram** [comment]
@ReinoutStevens I believe we are facing the same issue. We was using `onAuthStateChanged` then tried `authStateReady`.   Both have the same issue. Have you found a solution for this?

**Comment by ReinoutStevens** [comment]
No. We basically added a timeout, and when firebase doesnt resolve within that limit we show a popup to the user telling him to reload, and close other tabs of our website.

**Comment by jacobgram** [comment]
Thanks for sharing:

> We just found out that the behavior can be triggered by having multiple tabs open to the same site, and after several hours, opening a new tab will get stuck.

I was able to reproduce it.

**Comment by DellaBitta** [comment]
Hi @ReinoutStevens,

We're having trouble replicating this issue. Can you provide any more clues or insights on how best to reproduce it? Thanks!

**Comment by ReinoutStevens** [comment]
Hello,

We encounter it on https://www.crazygames.com
A reliable way to reproduce it is:

1. Open +- 10 tabs with our site while being logged in
2. Wait 25minutes
3. Open a new tab to crazygames.com: you should notice that now the header will display some spinners, which we display while we are waiting for the authstate callback to fire
4. Your open tabs will still work, but once you refresh them they will also be in the stuck state

I've attached a video that shows the problem

https://github.com/user-attachments/assets/099ceb17-f27a-44b8-b31a-34c5331c76c7

**Comment by DellaBitta** [comment]
Hi @ReinoutStevens,

This issue might stem from an IndexedDB issue in Safari's implementation of the API, but we're not entirely sure until we can reproduce it in-house and start to tinker with the implementation. Is there a chance you can make a reproducible isolated test case without the complexities of your entire web application?

We've tried to replicate the problem ourselves, starting with an empty project, but so far we haven't been able to produce the same error. Thanks!

**Comment by ReinoutStevens** [comment]
Hey google bot,

We are looking into making a minimal example, but it's going to take us some time to do so ( or to find time to begin with )
So no worries, we'll get back to you

**Comment by matt-wormley** [comment]
I believe this is the issue I am running into and I just filed a ticket with support. here is the minimal code that shows the issue. It does require multiple tabs, backgrounding the browser, and sleeping the phone (5 minutes). I can see hanging on one page that is only resolved after I reopen each of the tabs and then they all become unstuck.

When you see 1/null/##ms then there's no hang. With the hang the output is 0/null/?ms

```
import { useState, useEffect } from "react";
import { createRoot } from "react-dom/client";

import { initializeApp } from "firebase/app";
import { getAuth, onAuthStateChanged } from "firebase/auth";

const firebaseConfig = {...};

export const FirebaseApp = initializeApp(firebaseConfig);
export const FirebaseAuth = getAuth();

function App() {
  const [currentUser, setCurrentUser] = useState(FirebaseAuth.currentUser);
  const [count, setCount] = useState(-1);
  const [startTime, setStartTime] = useState<number | undefined>();
  const [callbackTime, setCallbackTime] = useState<number | undefined>();

  useEffect(() => {
    setStartTime(Date.now());
    let localCount = 0;
    setCount(localCount);
    return onAuthStateChanged(FirebaseAuth, (user) => {
      setCount(++localCount);
      setCallbackTime(Date.now());
      setCurrentUser(user);
    });
  }, [setCount, setCurrentUser]);

  return (
    <div>
      <p>{count}</p>
      <p>{!currentUser ? "null" : currentUser.isAnonymous ? "anon" : "user"}</p>
      <p>{(callbackTime && startTime && callbackTime - startTime) || "?"}ms</p>
    </div>
  );
}

createRoot(document.getElementById("root")!).render(<App />);
```

**Comment by matt-wormley** [comment]
It is looking like using local persistence works around this issue. Is this an issue with indexDb + ios?
```
initializeAuth(APP, {
        persistence: browserLocalPersistence,
        popupRedirectResolver: browserPopupRedirectResolver,
      });
```

**Comment by ReinoutStevens** [comment]
Thanks for the minimal example Matt
From our end:
* We switched to LS + indexeddb auth a couple weeks ago, which didnt fix things ( but made it so people did not have to login again )
* After that change, we switched to LS which has solved the problem

**Comment by gregwolanski** [comment]
> we switched to LS which has solved the problem

What do you mean by *LS*, @ReinoutStevens? 🙂

**Comment by gregwolanski** [comment]
Thank you for this clarification, @ReinoutStevens! 🫶 Have a wonderful day! 🙌

**Comment by flt-phy** [comment]
I have similar issue in Flutter Web, where initialization Firebase.initializeApp() takes forever, unless all other tabs (same app) are closed.  

We tried to switch to LOCAL storage, upgrade Firebase to latest versions, but it didn't help. Any idea would be so appreciated 🙏
